### PR TITLE
[FIX] mail: set full composer subject and recipients

### PR DESCRIPTION
**Pull request by xdu-odoo**
task-2291898
task-2291893
task-2291887

### CODE REVIEW

**Comment by xdu-odoo** [comment]
for recipients, how do you determine them ?

image

**Comment by alexkuhn** [comment]
> for recipients, how do you determine them ?
> 
> image

Related to suggested recipients, which is a missing feature (task-2280157). @phenix-factory is currently working on it.

**Comment by seb-odoo** [comment]
Also should add a test (tour), either now or with a new task.

**Comment by xdu-odoo** [comment]
> Also should add a test (tour), either now or with a new task.

For the tour I need a chatter on a document. Is there a way to ensure that an app is installed so that I can go to a form view with a chatter in it ?
We'll also need to do a master branch to introduce the new assets bundle as we did with the test in mail_entreprise I suppose

**Comment by seb-odoo** [comment]
> We'll also need to do a master branch to introduce the new assets bundle as we did with the test in mail_entreprise I suppose

For the assets it depends if the module on which we are going to add the tour already has that bundle `assets_tests` or not. But likely yes we'll need to do the trick indeed.

> For the tour I need a chatter on a record. Is there a way to ensure that an app is installed so that I can go to a form view with a chatter in it ?

You can check which views are extended in `mail` to add `oe_chatter`, which would definitely be the case for partner view.
The only question is what is the easiest way to reach it from the interface without other apps.
Probably settings > users > click on partner

Or the crappy way, just hardcode url as the base of the tour `/web#id=1&model=res.partner&view_type=form`

**Comment by xdu-odoo** [comment]
> > For the tour I need a chatter on a record. Is there a way to ensure that an app is installed so that I can go to a form view with a chatter in it ?
> 
> You can check which views are extended in `mail` to add `oe_chatter`, which would definitely be the case for partner view.
> The only question is what is the easiest way to reach it from the interface without other apps.
> Probably settings > users > click on partner

Managed to do it but how can I check the value on an input in a tour ? If I do $(input).val() in jQuery I have the right value but there is no value="" attribute so I can't select it that way in the tour. So I think I need a run() function but I'm not able to find in existing tour how I can check that and make the test fail if it's not. Do I throw an Exception ?

```
{
    content: "Check subject is autofilled",
    trigger: 'input[name="subject"]',
    run: (actions) => {
        // access trigger.val() and if empty raise Exception ?
    },
}
```

Tried with `trigger: 'input[name="subject"]:contains("Re: YourCompany, Marc Demo")'``` but didn't work as explained above

**Comment by seb-odoo** [comment]
Tour steps are based on DOM selector, so the best way for one step to be blocking/waiting on the result of the `run` is to make it so that the `run` is writing its result on the DOM (for example adding a data- attribute somewhere) and the next step is using that as trigger.

**Comment by xdu-odoo** [comment]
> Tour steps are based on DOM selector, so the best way for one step to be blocking/waiting on the result of the `run` is to make it so that the `run` is writing its result on the DOM (for example adding a data- attribute somewhere) and the next step is using that as trigger.

I found this kind of things (and it's working), what do you think ?

```js
tour._consume_tour(tour.running_tour, "Full composer does not have the right body content");
``

**Comment by seb-odoo** [comment]
I need to see the full example but calling a private method outside does not seem ideal.

**Comment by xdu-odoo** [comment]
> I need to see the full example but calling a private method outside does not seem ideal.

I'll push it so you can see, stock team use that and seems to be easier than writing something in the DOM IMO

**Comment by tde-banana-odoo** [comment]
I don't see the point of slowing down the runbot just to test that context is propagated ... this test is really not necessary.

File: addons/mail/static/src/components/composer/composer.js [hunk 1]
```diff
@@ -232,17 +232,19 @@ class Composer extends Component {
 
         const context = {
             // default_parent_id: this.id,
-            default_body: mailUtils.escapeAndCompactTextContent(this.composer.textInputContent),
             default_attachment_ids: attachmentIds,
+            default_body: mailUtils.escapeAndCompactTextContent(this.composer.textInputContent),
+            /* FIXME would need to use suggested_partners here task-2280157 */
             // default_partner_ids: partnerIds,
             default_is_log: this.composer.isLog,
+            default_use_template: true,
```
Comment (tde-banana-odoo) [inline addons/mail/static/src/components/composer/composer.js:240]: Why ? This should be set with a template_id set, or has no use.

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+const tour = require('web_tour.tour');
+
+/**
+ * This tour depends on data created by python test in charge of launching it.
+ * It is not intended to work when launched from interface. It is needed to test
+ * an action (action manager) which is not possible to test with QUnit.
+ * @see mail/tests/test_mail_full_composer.py
+ */
+tour.register('mail/static/tests/tours/mail_full_composer_test_tour.js', {
```
Comment (tde-banana-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:12]: According to me this is mainly noise; adding tests is ok, adding tests for a commented line only slows down runbot and add nothing. Moreover the "Re: blah" is already tested in python unit tests (much quicker, and along with other tests).

File: addons/mail/static/src/components/composer/composer.js (hunk 1 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/composer/composer.js:240]: Bad c/c I think, will remove

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js (hunk 2 above)
Comment (seb-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:12]: The full composer feature is considered important enough and it is currently not working at all. There are also related tasks 2291893 and 2291887 that should introduce more fixes to it (maybe to be done directly in current PR too).

It's not acceptable to have a feature like this one to be completely broken and runbot to be green.

Unfortunately introducing a tour (as opposed to writing a QUnit) is the only way we have to test this one due to the current implementation of the web client / action manager.

I hear your concern and I'm indeed aware of the slowness of tours, but I think in this case it's necessary / worth the time on runbot.

Comment (tde-banana-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:12]: Well actually this should not have passed review state if something like the composer is "currently nor working at all". I don't understand why a 2 lines fix about commented lines require a test, because there is no chance someone will remove those lines and it is not a corner case. Unless you plan to improve this test with other corner cases with other fixes, in which case it seems ok.

It fear adding too much tests for details may feel like we don't have to manually test anymore or ignore functional flows.

A global tour of composer is a good idea anyway, but I think it should be done in a real functional addon, like a complete flow of sale order confirmation + send by email + suggested recipients + email status .

Comment (xdu-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:12]: Just to be sure everyone is talking about the same thing : composer is globally working, SEB is talking about the "full size"  composer in modal which was not working and we didn't know about it. (If it was clear for everyone, sorry for that useless comment)

One of our goal (I think) is to have tours to test all mail functionalities as there are some features we can't totally cover with QUnit tests. The thing is that we are currently running out of time (like everybody I think) to add full tours. So this tour is very light for now but could be enriched later.

(@seb-odoo correct me if I'm wrong here)

Comment (seb-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:12]: Yes I think this is pretty much the idea. I would find great if we had a real tour with a complete flow such as the one of sales order as suggested, but we have to start somewhere, and doing directly the full tour is out of scope of current PR.

@xdu-odoo you can have a look at the other tasks I referenced above, maybe those 2 are worth immediately doing as part of this PR / this tour.

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    run: () => {
+        const dropdownElements = document.querySelectorAll('ul.ui-widget-content li');
+        const found = dropdownElements.reduce((acc, elem) => acc || elem.textContent.includes('Test template'), false);
+        if (!found) {
+            tour._consume_tour(
+                tour.running_tour,
+                "Full composer should contain at least one mail template"
+            );
+        }
+    },
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:64]: Issue is due to magic in tour manager that makes selector only look inside modal DOM when a modal is opened.
Problem is that autocomplete is outside of modal, so a special flag has to be added.

```suggestion
    content: "Check template is listed",
    trigger: '.ui-autocomplete .ui-menu-item a:contains(Test template)',
    in_modal: false,
```

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            tour._consume_tour(
+                tour.running_tour,
+                `Full composer should contain "blahblah" in body input (actual : ${bodyContent})`
+            );
+        }
+    },
+},{
+    content: "Open templates",
+    trigger: '.o_field_widget[name="template_id"] input',
+    run: 'click',
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:51]: no need to specify default action

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * It is not intended to work when launched from interface. It is needed to test
+ * an action (action manager) which is not possible to test with QUnit.
+ * @see mail/tests/test_mail_full_composer.py
+ */
+tour.register('mail/static/tests/tours/mail_full_composer_test_tour.js', {
+    test: true,
+}, [{
+    content: "Click on Send Message",
+    trigger: '.o_ChatterTopbar_buttonSendMessage',
+},{
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:17]: lint: space after `,`

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                `Full composer should have "Re: Test User" in subject input (actual : ${subjectValue})`
+            );
+        }
+    },
+},{
+    content: "Check composer content is kept",
+    trigger: '.oe_form_field[name="body"]',
+    run: () => {
+        const bodyContent = document.querySelector('.oe_form_field[name="body"] textarea').textContent;
+        if(!bodyContent.includes("blahblah")) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:41]: lint: space after `if`

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+},{
+    content: "Check subject is autofilled",
+    trigger: 'input[name="subject"]',
+    run: () => {
+        const subjectValue = document.querySelector('input[name="subject"]').value;
+        if (subjectValue !== "Re: Test User") {
+            tour._consume_tour(
+                tour.running_tour,
+                `Full composer should have "Re: Test User" in subject input (actual : ${subjectValue})`
+            );
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:33]: I think all this does is to` console.error` the message.

File: addons/mail/tests/test_mail_full_composer.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+@tagged('-at_install', 'post_install')
+class TestMailFullComposer(HttpCase):
+
+    def test_full_composer_tour(self):
+        self.env['mail.template'].create({
+            'name': 'Test template',
+            'partner_to': '${object.id}',
+            'lang': '${object.lang}',
+            'auto_delete': True,
+            'model_id': self.ref('base.model_res_partner')
```
Comment (seb-odoo) [inline addons/mail/tests/test_mail_full_composer.py:16]: lint: `,` at the end of multiline object

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js (hunk 4 above)
Comment (xdu-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:51]: I found that the default action when it's an input is to write "Test" in the input so I need to specify "click"

Comment (seb-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:51]: It's actually good that it writes that because that's part of the name of your template which is expected to be found.

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js (hunk 7 above)
Comment (xdu-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:33]: It makes the tour fail and print that message, yes

Comment (seb-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:33]: I mean just use console.error instead of calling an unrelated private method just to do console error.

File: addons/mail/static/src/components/composer/composer.js [hunk 9]
```diff
@@ -252,7 +253,14 @@ class Composer extends Component {
...
-        await this.env.bus.trigger('do-action', { action });
+        const options = {
+            on_close: () => {
+                if (this.composer.thread) {
+                    this.composer.thread.loadNewMessages();
+                }
+            },
+        };
+        await this.env.bus.trigger('do-action', { action, options });
         this.trigger('o-full-composer-opened');
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer/composer.js:264]: remove (unused)

**Comment by xdu-odoo** [comment]
> Also don't forget migration script (runbot's red because of that)

no migrations needed, we need to do a PR in master introducing the assets bundle and merge it before this PR

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * This tour depends on data created by python test in charge of launching it.
+ * It is not intended to work when launched from interface. It is needed to test
+ * an action (action manager) which is not possible to test with QUnit.
+ * @see mail/tests/test_mail_full_composer.py
+ */
+tour.register('mail/static/tests/tours/mail_full_composer_test_tour.js', {
+    test: true,
+},[{
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:14]: lint: space after `,` to apply in this whole file

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    content: "Check a template is listed",
+    in_modal: false,
+    trigger: '.ui-autocomplete .ui-menu-item a:contains("Test template")',
+    run: () => {},
+},{
+    content: "Send message",
+    trigger: '.o_mail_send',
+},{
+    content: "Check message is shown",
+    trigger: '.o_Message :contains("blahblah")',
```
Comment (seb-odoo) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:59]: The space before `:` is intended here? It feels like it shouldn't be needed, and if it is, you should directly target the proper children.

**Comment by xdu-odoo** [comment]
master PR for assets introduction : https://github.com/odoo/odoo/pull/54485

File: addons/mail/static/src/models/composer/composer.js [hunk 12]
```diff
@@ -81,6 +82,45 @@ function factory(dependencies) {
...
         }
 
+        /**
+         * Open the full composer modal.
+         */
+        async openFullComposer() {
+            const attachmentIds = this.attachments.map(attachment => attachment.res_id);
+
+            const context = {
+                // default_parent_id: this.id,
```
[the remark this answers is not in the document]
Comment (xdu-odoo) [reply addons/mail/static/src/models/composer/composer.js:92]: in wizard model "mail.composer.message"
```
    parent_id = fields.Many2one(
        'mail.message', 'Parent Message', index=True, ondelete='set null',
        help="Initial thread message.")
```

I think this is when you reply to a message or something like that ?

File: addons/mail/static/src/models/composer/composer.js [hunk 13]
```diff
@@ -81,6 +82,45 @@ function factory(dependencies) {
...
+                view_mode: 'form',
+                views: [[false, 'form']],
+                target: 'new',
+                context: context,
+            };
+            const options = {
+                on_close: () => {
+                    if (this.thread) {
+                        this.thread.loadNewMessages();
+                    }
```
[the remark this answers is not in the document]
Comment (xdu-odoo) [reply addons/mail/static/src/models/composer/composer.js:118]: I didn't understand ... I should do `this.constructor.get(this).thread.loadNewMessages()` ?

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/composer/composer.js:118]: `if (this.constructor.get(this) && this.thread) {`

File: addons/mail/static/src/models/composer/composer.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/models/composer/composer.js:92]: Yes, I don't really know more than you.

Old code apparently set this value to `self.id` where self was the `ChatterComposer`. Except I don't see anywhere in old code in `ChatterComposer` or `BasicComposer` where `this.id` is assigned, so maybe it was always unset in the context of the chatter. I just made an attempt in 13.3 and it confirms this (null is sent).

@alexkuhn any idea?

[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/models/composer/composer.js:92]: No idea, looks like dead code.

[the remark this answers is not in the document]
Comment (xdu-odoo) [reply addons/mail/static/src/models/composer/composer.js:92]: So I just remove the line ?

File: addons/mail/tests/test_mail_full_composer.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.tests.common import tagged, HttpCase
+
+
+@tagged('-at_install', 'post_install', 'blah')
```
Comment (seb-odoo) [inline addons/mail/tests/test_mail_full_composer.py:7]: still have to remove blah from here

File: addons/mail/static/src/models/composer/composer.js [hunk 15]
```diff
@@ -81,6 +82,44 @@ function factory(dependencies) {
...
+                view_mode: 'form',
+                views: [[false, 'form']],
+                target: 'new',
+                context: context,
+            };
+            const options = {
+                on_close: () => {
+                    if (this.constructor.get(this) && this.thread) {
+                        this.thread.loadNewMessages();
+                    }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:117]: Looks almost fine now.

Just one small issue after testing, I noticed that the text of the composer is kept in store even after the message is posted through modal. This is because wizard doesn't go through `postMessage` that does `_reset`.

Also "small" composer content is not sync'ed with the full composer in case of close, it still has the old content.
Ideally both should be sync'ed, but this is out of scope/almost impossible since an action is used.

I think an acceptable alternative for now would be to call `_reset`, either when we open or when we close the action, cc @alexkuhn?

Comment (seb-odoo) [reply addons/mail/static/src/models/composer/composer.js:117]: Long term maybe the "small composer" should also instantiate a `mail.compose.message` and both should be server sync with it?

Comment (alexkuhn) [reply addons/mail/static/src/models/composer/composer.js:117]: > I think an acceptable alternative for now would be to call `_reset`, either when we open or when we close the action, cc @alexkuhn?

Yes

File: addons/mail/static/src/models/composer/composer.js [hunk 16]
```diff
@@ -81,6 +82,47 @@ function factory(dependencies) {
...
+         */
+        async openFullComposer() {
+            const attachmentIds = this.attachments.map(attachment => attachment.res_id);
+
+            const context = {
+                default_attachment_ids: attachmentIds,
+                default_body: mailUtils.escapeAndCompactTextContent(this.textInputContent),
+                /* FIXME would need to use suggested_partners here task-2280157 */
+                // default_partner_ids: partnerIds,
+                default_is_log: this.isLog,
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:96]: alphabetical order (`default_partner_ids` after `default_is_log`)

File: addons/mail/static/src/models/composer/composer.js [hunk 17]
```diff
@@ -81,6 +82,47 @@ function factory(dependencies) {
...
+                mail_post_autofollow: true,
+            };
+
+            if (this.thread) {
+                context.default_model = this.thread.model;
+                context.default_res_id = this.thread.id;
+            }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:103]: The guarding is weird. `mail.composer` is modeled as a part of `mail.thread`, so functionally a `mail.composer` doesn't exist without `mail.thread` (minus accidental-and-critical-and-to-be-fixed time frames like teardown in tests, but this code is not called during this time).

File: addons/mail/static/src/models/composer/composer.js [hunk 18]
```diff
@@ -81,6 +82,47 @@ function factory(dependencies) {
...
+                context: context,
+            };
+            const options = {
+                on_close: () => {
+                    if (this.constructor.get(this)) {
+                        this._reset();
+                        if (this.thread) {
+                            this.thread.loadNewMessages();
+                        }
+                    }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:120]: just pre-emptive return (readability)

File: addons/mail/static/src/models/composer/composer.js [hunk 19]
```diff
@@ -81,6 +82,47 @@ function factory(dependencies) {
...
+                target: 'new',
+                context: context,
+            };
+            const options = {
+                on_close: () => {
+                    if (this.constructor.get(this)) {
+                        this._reset();
+                        if (this.thread) {
+                            this.thread.loadNewMessages();
+                        }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:119]: Same remark on weird guarding.

File: addons/mail/static/tests/tours/mail_full_composer_test_tour.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    },
+}, {
+    content: "Check composer content is kept",
+    trigger: '.oe_form_field[name="body"]',
+    run: () => {
+        const bodyContent = document.querySelector('.oe_form_field[name="body"] textarea').textContent;
+        if (!bodyContent.includes("blahblah")) {
+            console.error(
+                `Full composer should contain "blahblah" in body input (actual : ${bodyContent})`
```
Comment (alexkuhn) [inline addons/mail/static/tests/tours/mail_full_composer_test_tour.js:42]: Would just clarify text contents are recovered from small composer

Comment (xdu-odoo) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js:42]: text content is recovered but subject is not, it's built on record name I think

Comment (alexkuhn) [reply addons/mail/static/tests/tours/mail_full_composer_test_tour.js]: ok

File: addons/mail/static/src/models/composer/composer.js [hunk 21]
```diff
@@ -81,6 +82,43 @@ function factory(dependencies) {
...
+        /**
+         * Open the full composer modal.
+         */
+        async openFullComposer() {
+            const attachmentIds = this.attachments.map(attachment => attachment.res_id);
+
+            const context = {
+                default_attachment_ids: attachmentIds,
+                default_body: mailUtils.escapeAndCompactTextContent(this.textInputContent),
+                /* FIXME would need to use suggested_partners here task-2280157 */
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:94]: comment should be above commented `default_partner_ids: partnerIds,` line
